core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -1281,13 +1389,323 @@ public String toStringResponseOnly(Function<? super ResponseHeaders, ?> headersS
+        }
+
+        @Override
+        public boolean isAvailable(RequestLogProperty property) {
+            return true;
+        }
+
+        @Override
+        public RequestLog partial() {
+            return this;
+        }
+
+        @Override
+        public CompletableFuture<RequestLog> whenComplete() {
+            if (completedFuture == null) {
+                completedFuture = UnmodifiableFuture.completedFuture(this);
+            }
+            return completedFuture;
+        }
+
+        @Override
+        public CompletableFuture<RequestOnlyLog> whenRequestComplete() {
+            // OK to cast because the future is unmodifiable.
+            @SuppressWarnings("unchecked")
+            final CompletableFuture<RequestOnlyLog> cast =
+                    (CompletableFuture<RequestOnlyLog>) (CompletableFuture<?>) whenComplete();
+            return cast;
+        }
+
+        @Override
+        public CompletableFuture<RequestLog> whenAvailable(RequestLogProperty property) {
+            return whenComplete();
+        }
+
+        @Override
+        public CompletableFuture<RequestLog> whenAvailable(RequestLogProperty... properties) {
+            return whenComplete();
+        }
+
+        @Override
+        public CompletableFuture<RequestLog> whenAvailable(Iterable<RequestLogProperty> properties) {
+            return whenComplete();
+        }
+
+        @Override
+        public RequestLog ensureComplete() {
+            return this;
+        }
+
+        @Override
+        public RequestOnlyLog ensureRequestComplete() {
+            return this;
+        }
+
+        @Override
+        public RequestLog ensureAvailable(RequestLogProperty property) {
+            return this;
+        }
+
+        @Override
+        public RequestLog ensureAvailable(RequestLogProperty... properties) {
+            return this;
+        }
+
+        @Override
+        public RequestLog ensureAvailable(Iterable<RequestLogProperty> properties) {
+            return this;
+        }
+
+        @Override
+        public int availabilityStamp() {
+            return RequestLogProperty.FLAGS_ALL_COMPLETE;
+        }
+
+        @Override
+        public RequestContext context() {
+            return ctx;
+        }
+
+        @Override
+        public List<RequestLogAccess> children() {
+            return DefaultRequestLog.this.children();
+        }
+
+        @Override
+        public long requestStartTimeMicros() {
+            return requestStartTimeMicros;
+        }
+
+        @Override
+        public long requestStartTimeMillis() {
+            return TimeUnit.MICROSECONDS.toMillis(requestStartTimeMicros);
+        }
+
+        @Override
+        public long requestStartTimeNanos() {
+            return requestStartTimeNanos;
+        }
+
+        @Override
+        public Long requestFirstBytesTransferredTimeNanos() {
+            return requestFirstBytesTransferredTimeNanosSet ? requestFirstBytesTransferredTimeNanos : null;
```

Comment:
If we do `DefaultRequestLog.this.isAvailable(RequestLogProperty.REQUEST_FIRST_BYTES_TRANSFERRED_TIME)`
we can remove `requestFirstBytesTransferredTimeNanosSet` completely.

Reply:
Yeah, but the point of this change is to make all properties behave in a consistent way.

When a certain property is not available, it means we do not know its value yet and we may know it later.

When available, it means we do know the value and it will not change.

Previously, `{request,response}FirstBytesTransferredTime` had no distinction between 'not written so far but may be written in the future' and 'not written so far and will never be written' unlike other properties. 

A user was able to tell by checking whether the request is complete, but it is indirect and a user was not able to write a simple listener like this:

```java
log.addListener(log -> {
    // Do this if written. Do that if not written.
}, REQUEST_FIRST_BYTES_TRANSFERRED);
```

.. because the callback will not always be invoked.

The new API guaranteed the callback will be invoked, which is less surprising:

```java
log.whenAvailable(REQUEST_FIRST_BYTES_TRANSFERRED_TIME).thenAccept(log -> {
    // Do this if written, Do that if not written.
});
```